### ISSUE TITLE
Question: Option for redelivering unacknowledged RabbitMQ messages

### ISSUE THREAD

**Issue by BentCoder** [php-amqplib/RabbitMqBundle#328]
Hi,

Is there a "Redelivering unacknowledged messages" option that we can use?

Thanks for the great work for creating this bundle for Symfony users! I hope someone with appropriate knowledge takes over it.

**My unacknowledged messages**

```
$ rabbitmqctl list_queues name messages_ready messages_unacknowledged
Listing queues ...
order_create_qu 0   2
```

**My Publisher**

```
use OldSound\RabbitMqBundle\RabbitMq\ProducerInterface;

class OrderCreateProducer
{
    private $producer;

    public function __construct(ProducerInterface $producer)
    {
        $this->producer = $producer;
    }

    public function produce()
    {
        $message = [id' => 123];

        $this->producer->publish(json_encode($message));
    }
}
```

**My Consumer**

```
use Exception;
use OldSound\RabbitMqBundle\RabbitMq\ConsumerInterface;
use PhpAmqpLib\Message\AMQPMessage;

class OrderCreateConsumer implements ConsumerInterface
{
    public function execute(AMQPMessage $message)
    {
        try {
            // Do something with $message->body;
        } catch (Exception $e) {
            return false;
        }
    }
}
```

**Comment by cmodijk** [comment]
Instead of `return false` you could use `return ConsumerInterface:: MSG_REJECT_REQUEUE` which forces the message to be redelivered the only problem is that because RabbitMQ is a fifo (first in first out) system the message gets redelivered as the first message in line and if it keeps failing al messages behind that message are not getting processed (with 1 consumer). 

The solution we use is we `return ConsumerInterface::MSG_REJECT` which means it failed and don't requeue but because RabbitMQ now knows it failed you could setup a dead letter exchange to keep failed messages. I have an outstanding PR to the `SimpleBus/RabbitMQBundleBridge` describing how we handle failure. Here is the document https://github.com/cmodijk/RabbitMQBundleBridge/blob/patch-3/doc/dead_letter_exchange.md hope this helps.

**Comment by BentCoder** [comment]
@cmodijk Thank you for enlightening me and it has been helpful. What I want to ask is (based on my searches) this bundle (php-amqplib/RabbitMqBundle) doesn't support Dead Letter Exchange yet, doesn't it? I'm asking because I haven't seen any examples and configurations regarding to it.

**Comment by cmodijk** [comment]
@BentCoder Yeah it's possible with the bundle an example configuration can be found in my pull request for RabbitMQBundleBridge it's not specific to the RabbitMQBundleBridge. 

```
# in config.yml
old_sound_rabbit_mq:
    # don't forget to provide the connection details
    ...
    producers:
        ...
        your_job:
            connection:       default
            exchange_options: { name: 'your_job', type: direct }
    consumers:
        ...
        your_job_fallback:
            connection:       default
            exchange_options: { name: 'your_job_fallback', type: direct }
            queue_options:    { name: 'your_job_fallback' }
            callback:         yourcallback
        your_job:
            connection:       default
            exchange_options: { name: 'your_job', type: direct }
            queue_options:    { name: 'your_job', arguments: {'x-dead-letter-exchange': ['S', 'your_job_fallback']} }
            callback:         yourcallback
```

and this is the example for doing an auto retry every hour 

```
# in config.yml
old_sound_rabbit_mq:
    # don't forget to provide the connection details
    ...
    producers:
        ...
        your_job:
            connection:       default
            exchange_options: { name: 'your_job', type: direct }
    consumers:
        ...
        your_job_fallback:
            connection:       default
            exchange_options: { name: 'your_job_fallback', type: direct }
            queue_options:    { name: 'your_job_fallback', arguments: {'x-dead-letter-exchange': ['S', 'your_job'], 'x-message-ttl': ['I', 3600000]} }
            callback:         yourcallback
        your_job:
            connection:       default
            exchange_options: { name: 'your_job', type: direct }
            queue_options:    { name: 'your_job', arguments: {'x-dead-letter-exchange': ['S', 'your_job_fallback']} }
            callback:         yourcallback
```

https://github.com/cmodijk/RabbitMQBundleBridge/blob/patch-3/doc/dead_letter_exchange.md

**Comment by BentCoder** [comment]
@cmodijk I did it like yours but the rejected messages are not going to dead letter exchange and its queue. I removed `x-dead-letter-routing-key` and no change as well.

```
old_sound_rabbit_mq:
    connections:
        default:
            host:     %rabbitmq.host%
            port:     %rabbitmq.port%
            user:     %rabbitmq.user%
            password: %rabbitmq.pswd%
            vhost:    /
            lazy:     true
    producers:
        order_create:
            connection:       default
            exchange_options: { name: order_create_ex, type: direct }
    consumers:
        order_create:
            connection:       default
            exchange_options: { name: order_create_ex, type: direct }
            queue_options:    { name: order_create_qu, arguments: { 'x-dead-letter-exchange': ['S', 'order_create_dead_letter'], 'x-dead-letter-routing-key': ['S', 'order_create_dead_letter_qu'] } }
            callback:         application_frontend.consumer.order_create
        order_create_dead_letter:
            connection:       default
            exchange_options: { name: order_create_dead_letter_ex, type: direct }
            queue_options:    { name: order_create_dead_letter_qu }
            callback:         application_frontend.consumer.order_create_dead_letter
```

**Comment by geekdevs** [comment]
Note that fallback queue is not created automatically unless you publish a message to it directly.
You should run `bin/console rabbitmq:setup-fabric` so that all queues/exchanges from your symfony config are synchronized with rabbitmq server.

**Comment by BentCoder** [comment]
I had a long break and decided to take on this issue again. The problem was simple.

Change `'x-dead-letter-exchange': ['S', 'order_create_dead_letter']` to `'x-dead-letter-exchange': ['S', 'order_create_dead_letter_ex']` so it matches the exchange name of the actual DLX.

That's it!

**Comment by fredleger** [comment]
@BentCoder @cmodijk i know this is quite old but digged into this today. Did you manage to make it work ? my tries heres seems to show that this is not working anymore and anything you could do (even returning `ConsumerInterface::MSG_REJECT`) will requeue the msg.

Am i right on this assumption ?

**Comment by BentCoder** [comment]
@fredleger I remember solving it but cannot remember how. I think I used [DLX](http://www.inanzzz.com/index.php/post/ovbz/using-rabbitmq-dead-letter-exchange-feature-to-store-rejected-or-corrupted-messages-within-symfony) or something. Have a look at [this](http://www.inanzzz.com/index.php/post/4b5a/the-proper-ways-of-handling-errors-in-symfony-rabbitmq-consumer). It tells you all about the options. Having said that, I would encourage you look into https://github.com/symfony/messenger for future development on Rabbit. It is very simple and in constant development. Although this bundle has helped me a lot in the past, it is in a state of limbo. I stopped using it permanently.
